Project planning building

dcoutts opened the pull request:
Ready for review.

Ignore the first 8 patches, they're inherited from PR #3226. It's just the last 4.

It consists of two major phases, planning and building (each of which consists of several passes mostly involving the ElaboratedInstallPlan data structure).

23Skidoo commented on cabal-install/Distribution/Client/InstallPlan.hs:
```diff
@@ -486,6 +490,69 @@ checkConfiguredPackage (Failed     _ _) = Nothing
+                  $ planIndex plan
+    }
+
+mapPreservingGraph :: (HasUnitId ipkg,
```
This needs a Haddock comment.

23Skidoo commented:
/cc @ezyang

23Skidoo commented:
LGTM in general. Some top-level definitions are missing Haddock strings, but otherwise the code is very well commented, so I think the PR can be merged as it is.

ezyang commented:
There are some typos in the commit messages, e.g. "hasing" in https://github.com/haskell/cabal/pull/3227/commits/2cc56a9e2ff2869f6879b5f8195a3c2afac3d1cc

ezyang commented:
I'd quite like it if the hashes were base 62 encoded. Will save space.

ezyang commented on cabal-install/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -0,0 +1,1257 @@
+          newPackageFileMonitor distDirLayout (packageId pkg)
+
+
+traverseInstallPlan :: forall m ipkg srcpkg iresult ifailure b.
```
Needs a comment saying that it feeds in the produced b's of your dependencies. Not really a "traverse" in the traversable sense.

ezyang commented on cabal-install/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -0,0 +1,1257 @@
+      | InstallPlan.Configured pkg
+          <- InstallPlan.reverseTopologicalOrder installPlan
+      , let ipkgid = installedPackageId pkg
+      , BuildStatusUpToDate mipkg buildSuccess <- [pkgsBuildStatus Map.! ipkgid]
```
Partial?!

dcoutts replied:
You mean the `BuildStatusUpToDate` matching or the Map.! ?

The pkgsBuildStatus Map covers every package in the plan, it's an internal error if not. The pattern match on BuildStatusUpToDate is not partial, it's a filter. It is only those packages where the build status says we are up to date that we want to change.

ezyang replied:
Map.!. I'm sure it's an internal error but it's kinder to future debuggers if the invariant breaks to give a specific error message which can be corroborated with the code that assumes this.

dcoutts replied:
Yes ok fair enough. I'll use a `let Just blah = Map.lookup` style so we get a clear exception with a location.

ezyang commented on cabal-install/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -0,0 +1,1257 @@
+-- for other changes. This is the purpose of 'updatePackageConfigFileMonitor'
+-- and 'updatePackageBuildFileMonitor'.
+--
+data PackageFileMonitor = PackageFileMonitor {
```
I couldn't remember what the () parameters mean. Had to go to FileMonitor and remind myself that it's a map from key to value, so map to () is a set and map from () is a singleton. Maybe we should alias.

ezyang commented on cabal-install/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -0,0 +1,1257 @@
+packageFileMonitorKeyValues pkg =
+    (pkgconfig, buildComponents)
+  where
+    pkgconfig = pkg {
```
I don't really understand what this does. Why do we nub out the targets?

dcoutts replied:
In a sense it's a slight hack. The `ElaboratedConfiguredPackage` consists mostly of inputs to the package configure step, but there's also a few fields for which targets to build. These do not affect whether or not we need to re-run the configure step. And then we track which set of components we've built separately.

If that's satisfactory I can add that explanation as a comment. The other approach would be to split the `ElaboratedConfiguredPackage` to keep the configure step vs build step info separate in the first place, but in other places it's rather convenient to keep them together. I think my view is that on balance it's easier this way, but it certainly deserves explanation.

ezyang replied:
OK explanation seems fine.

ezyang commented on cabal-install/Distribution/Client/ProjectBuilding.hs:
```diff
@@ -0,0 +1,1257 @@
+        | otherwise -> do
+            buildChanged  <- checkFileMonitorChanged
+                               pkgFileMonitorBuild srcdir buildComponents
+            regChanged    <- checkFileMonitorChanged
```
"In fact, you don't need to probe the reg unless monitor is unchanged, but then the case blocks nest too much." There ought to be some combinator which solves this problem, I think.

ezyang commented:
OK, looks good to me. (Just some annoying CR comments which could be ignored...)

dcoutts commented:
@ezyang I've just been having a look at base62. It doesn't look very easy to do. I see we have some code in Cabal already but that's for when you're starting with an Integer, whereas we are starting with a bytestring. We could use a base64 variant, for which there's easy code we can pinch, but we'd have to pick two valid file chars that we're happy with using, e.g. '%', '@', '_', '+' or '~'.

ezyang commented:
I was going to suggest you use `fingerprintData` in `GHC.Fingerprint` but then I remembered that you are using a different hash algorithm here. Maybe base64 is OK, but you will run into difficulty picking extra characters which are also valid IPIDs: in fact, your only choices for extra characters are these three: "-_.". Another possibility is to do "base62 but a little lossy": basically, every 32-bits is encoded as 11 base 62 characters. I think this will be one extra character than the optimal base 62 encoding... which is not bad, TBH.

dcoutts commented:
@ezyang I'd be ok with base64 using '.' and '_'. What do you reckon?

dcoutts commented:
@ezyang pushed a patch to add it. Let me know what you think.

dcoutts commented:
@ezyang I've had it pointed out to me that base64 is no good because of case insensitive file systems. We can only use a-z or A-Z not both. So base32 seems better. It's also what nix uses apparently (or at least it uses some variant of base32 that uses only lower case letters and numbers).

dcoutts commented:
@ezyang Ok, so I've implemented base32, but on reflection I'm not sure it's really worth it. Here's the balance:

Positives:
* Saves 12 characters, or 15%. With base16 the average package id length is 81, 64 for the sha256 hash and an average of 17 for the name + version. So with base32 the average would be 69 which is a saving of just under 15%.

Negatives:
 * Non-standard hash format. I implemented Nix's base32 variant which is non-standard. Standard RFC base32 is not suitable due to the use of padding. Common unix tools don't do base32.

With base16 we get this nice simple property:

    $ cd ~/.cabal/store/ghc-7.10.1/
    $ sha256sum ansi-terminal-0.6.2.2-6e0ddba7fd2190f6efa191bf63ed0ba81b071c337d1663b3e0e3984be7c16159/cabal-hash.txt
    6e0ddba7fd2190f6efa191bf63ed0ba81b071c337d1663b3e0e3984be7c16159  ansi-terminal-0.6.2.2-6e0ddba7fd2190f6efa191bf63ed0ba81b071c337d1663b3e0e3984be7c16159/cabal-hash.txt

That is, using the standard unix tool, hashing the text file with the package metadata gives us exactly the hash in the package id.

ezyang commented:
Oof, case-insensitive file systems are the dumps. I agree that there is not much point going from base16 to base32 (base64 helps a bit more.) I suppose this also means that the current base62 code in Cabal is probably being misused, because I'm pretty sure these IDs are being used for file system paths. (Of course, the collision possibility is still fairly rare, so I don't think it is affecting many people.)

One thing that is very annoying about all this is these SHA256 IDs will make there way into the GHC compilation process, and, e.g., be used to assign symbol names to generated code. I had previously come to terms with MD5/SHA1 sized hashes being embedded in, but SHA256 is going to make things even longer (another 50%). :/

dcoutts commented:
@ezyang it's not essential (afaik) that the symbol names be the full package manager assigned package ids. We can probably prune those.

Ok, so I'll rebase and merge this patchset without the base64/32 hash changes.

 @23Skidoo @ezyang  Any objections (e.g. comments I didn't address), say so now.

23Skidoo commented:
Let's merge!

mitchty commented on cabal-install/cabal-install.cabal:
```diff
@@ -214,12 +217,15 @@ executable cabal
     -- NOTE: when updating build-depends, don't forget to update version regexps
     -- in bootstrap.sh.
     build-depends:
+        async      >= 2.0      && < 3,
```
Thinking this and bitable and cryptohash should be in bootstrap.sh as well no?

23Skidoo replied:
Added `async` (see https://github.com/haskell/cabal/pull/3244). `cryptohash` and `byteable` are already there, since `hackage-security` depends on them.

23Skidoo commented on cabal-install/cabal-install.cabal:
```diff
@@ -214,12 +217,15 @@ executable cabal
         bytestring >= 0.9      && < 1,
         Cabal      >= 1.24     && < 1.25,
         containers >= 0.4      && < 0.6,
+        cryptohash >= 0.11     && < 0.12,
```
@dcoutts BTW, `cryptohash` has been deprecated in favour of `cryptonite`, and versions starting with 0.11.8 pull in the whole of `cryptonite`. So we probably should set the upper bound for `cryptohash` at 0.11.8.

23Skidoo replied:
See also https://github.com/well-typed/hackage-security/issues/152.

ezyang commented:
@dcoutts I wonder, can't we use the base62/base64 code for case-sensitive filesystems, and use base16 for case-insensitive? I'm still a bit loathe to give up what is essentially a very nice space saving on the IDs.
